galileo/src/layer/vector_tile_layer/expressions.rs

```diff
@@ -0,0 +1,443 @@
+pub struct StepValue<T> {
+    resolution: f64,
+    step_value: T,
+}
+
+/// Type used to define expressions for interpolation
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct InterpolateExpression<T> {
+    interpolation_type: Interpolation,
+    interpolation_args: InterpolationArgs<T>,
+}
+/// Type used to define steps
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct StepExpression<T> {
+    default_value: T,
+    /// Each stop value maps the resolution to the T type
+    /// If, the current resolution is greater than step resolution
+    /// the value T maps to the T value where the step resolution is
+    /// less than that of current resolution.
+    step_values: Option<Vec<StepValue<T>>>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct ResolutionValueRange<T> {
+    max_resolution: f64,
+    min_resolution: f64,
+    start_value: Option<T>,
+    end_value: Option<T>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct ValueStep<T> {
+    resolution: f64,
+    value: T,
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum Interpolation {
+    /// Linear interpolation type with base 1
+    Linear,
+    /// Exponential interpolation with variable base
+    Exponential,
+    /// Cubic Bezier interpolation
+    Cubic,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub enum StyleValue<T: Ord> {
+    /// Simple values like using Color::BLACK
+    Simple(T),
+    /// Interpolate expression
+    Interpolate(InterpolateExpression<T>),
+    /// Step expression
+    Steps(StepExpression<T>),
+}
+
+impl Color {
+    /// Evaluates value from simple color
+    pub fn get_value(&self) -> Result<Color, GalileoError> {
+        Ok(*self)
+    }
+}
+
+impl InterpolateExpression<Color> {
+    /// Evaluates value by interpolating color values
+    pub fn get_value(&self, current_resolution: f64) -> Option<Color> {
+        let resolution_value_range: Option<ResolutionValueRange<Color>> =
+            self.get_resolution_value_range(current_resolution);
+
+        if let Some(resolution_value_range) = resolution_value_range {
+            if current_resolution == resolution_value_range.max_resolution {
+                return resolution_value_range.start_value;
+            } else if current_resolution == resolution_value_range.min_resolution {
+                return resolution_value_range.end_value;
+            } else {
+                match self.interpolation_type {
+                    Interpolation::Linear => {
+                        let interpolated_color: Color = Self::lerp(
+                            resolution_value_range.start_value?,
+                            resolution_value_range.end_value?,
+                            Some(resolution_value_range.min_resolution),
+                            Some(resolution_value_range.max_resolution),
```

Comment:
Can't we just give the range to the function instead of all it's fields separately?